### DISCUSSION TITLE
Debugging corrupted PDFs turns out to be a problem with Quarto after 1.2.233 with latexmk

### DISCUSSION THREAD

**Discussion by fuhrmanator** [quarto-dev/quarto-cli#3359] [Q&A] [answered]
I spent several hours trying to track down why my CI for the Quarto book project I have (running on ubuntu) wasn't building a valid PDF (there was no error in the build). I had made some big changes to my book project and hadn't verified the PDF on the web, and thought it was I who "broke" things with ubuntu (I actually didn't realize the CI was running the release candidate builds). The PDF was coming out corrupt. I tried to remove all the changes I'd made in the past week, but couldn't get a MWE; the simplest book (hello world preface) was still a corrupt PDF.

On my Windows station everything is OK. I'm using Quarto 1.2.233, and the PDFs are fine. So, when I realized the CI builds were moving ahead of my windows version, I set them to 1.2.233 also, and now everything is OK.

I suspect it has something to do with `pdf-engine: latexmk`, since on a separate laptop I can get a Quarto project to render good PDFs if I don't use `latexmk` with the latest (1.2.269). It looks like some kind of regression (not sure when this started, sorry).

**Comment by rgaiacs** [comment]
A minimal working example to help debugging:

``````
---
title: "MEW"
keep-md: true
---

## Quarto

Quarto enables you to weave together content and executable code into a finished document.
To learn more about Quarto see <https://quarto.org>.

```{r}
1 + 1
```
``````

I render the above document using `quarto render --to pdf 3359.qmd` and the output is

```
processing file: 3359.qmd
  |.......................                                               |  33%
  ordinary text without R code

  |...............................................                       |  67%
label: unnamed-chunk-1
  |......................................................................| 100%
  ordinary text without R code


output file: 3359.knit.md

pandoc 3359.qmd
  to: latex
  output-file: 3359.tex
  standalone: true
  pdf-engine: xelatex
  variables:
    graphics: true
    tables: true
  default-image-extension: pdf
  
metadata
  documentclass: scrartcl
  classoption:
    - DIV=11
    - numbers=noendperiod
  papersize: letter
  header-includes:
    - '\KOMAoption{captions}{tableheading}'
  block-headings: true
  title: MEW
  
running xelatex - 1
  This is XeTeX, Version 3.141592653-2.6-0.999993 (TeX Live 2022/dev/Debian) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  
running xelatex - 2
  This is XeTeX, Version 3.141592653-2.6-0.999993 (TeX Live 2022/dev/Debian) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  

Output created: 3359.pdf
```

The intermediate Markdown files is as expected:

``````
---
title: "MEW"
keep-md: true
---



## Quarto

Quarto enables you to weave together content and executable code into a finished document.
To learn more about Quarto see <https://quarto.org>.


::: {.cell}

```{.r .cell-code}
1 + 1
```

::: {.cell-output .cell-output-stdout}
```
[1] 2
```
:::
:::
``````

but the PDF has duplicated content:

![Screenshot from 2022-11-17 11-13-56](https://user-images.githubusercontent.com/1506457/202345985-cdba84eb-0504-48ec-89c7-f8b53ffb94c2.png)

This duplication of content might be the source of the corrupted PDF.

## Software Stack

```
$ quarto check

[✓] Checking Quarto installation......OK
      Version: 1.2.269
      Path: /opt/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.9.10 (Conda)
      Path: /home/raniere/mambaforge/bin/python
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with conda install jupyter

[✓] Checking R installation...........OK
      Version: 4.2.1
      Path: /usr/lib/R
      LibPaths:
        - /home/raniere/R/x86_64-pc-linux-gnu-library/4.2
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      rmarkdown: 2.16

[✓] Checking Knitr engine render......OK
$ xelatex --version
XeTeX 3.141592653-2.6-0.999993 (TeX Live 2022/dev/Debian)
kpathsea version 6.3.4/dev
Copyright 2021 SIL International, Jonathan Kew and Khaled Hosny.
There is NO warranty.  Redistribution of this software is
covered by the terms of both the XeTeX copyright and
the Lesser GNU General Public License.
For more information about these matters, see the file
named COPYING and the XeTeX source.
Primary author of XeTeX: Jonathan Kew.
Compiled with ICU version 70.1; using 70.1
Compiled with zlib version 1.2.11; using 1.2.11
Compiled with FreeType2 version 2.11.1; using 2.11.1
Compiled with Graphite2 version 1.3.14; using 1.3.14
Compiled with HarfBuzz version 2.7.4; using 2.7.4
Compiled with libpng version 1.6.37; using 1.6.37
Compiled with pplib version v2.05 less toxic i hope
Compiled with fontconfig version 2.13.1; using 2.13.1
$ latexmk --version

Latexmk, John Collins, 20 November 2021. Version 4.76
```

**Reply by fuhrmanator** [reply]
Well done! I confirm the size of the files practically doubled during the change.

**Reply by dragonstyle** [reply]
We did make one change with regards to `latexmk` pdf-engine in 1.2, so it might be related to that. I've created an issue to track this and will have a look shortly.

**Reply by dragonstyle** [reply]
I can't reproduce the issue using the MWE example above (run locally, that produces the expected PDF with valid unduplicated content for me). When specifying the `latexmk` `pdf-engine`, however I can reproduce the issue. I am working on a fix.

**Reply by fuhrmanator** [reply]
Thanks @rgaiacs for making the MWE

**Reply by fuhrmanator** [reply]
On the separate Windows laptop using `latexmk`, the result of my book project (not minimal) is a PDF that opens without erroring. However, the content appears blank (there are many pages, but they are blank).

**Reply by dragonstyle** [reply]
Ok I've confirmed and fixed the issue in our latest 1.3 prerelease build. The issue / fix is:

> This issue was caused by a change in 1.2 to the way we handled pdf generation when the engine was latexmk. Normally, we use Pandoc to generate LaTeX, then run postprocessors on LaTeX, then generate a PDF from the completed LaTeX.
>
> But when the engine was set of latexmk, we were having Pandoc make the PDF, then running the post processors on the PDF (not the LaTeX), which was corrupting it.
>
> Starting in Quarto 1.3, we will always have Pandoc generate LaTeX and will use the appropriate engine to actually render the PDF (latexmk will mean to bypass the Quarto PDF generation loop and use the latexmk loop instead).

I was able to confirm that all rendered fine using your repo (to support your book with plantUML and remote images, I also added support for SVG conversion and remote image retrieval as well). If you have time to try the latest version and report back, it would be appreciated!

**Reply by rgaiacs** [reply]
@dragonstyle Thanks for the explanation and the fix. Can you explain a little more the difference between `latexmk` and `quarto` PDF generation loop? My understanding of LaTeX ecosystem is that to get the cross reference working, the user must run the LaTeX compiler three times and extra processing programs. `latexmk` automates the calls to LaTeX compiler and extra processing programs.

**Reply by fuhrmanator** [reply]
Today I ran Quarto 1.3.24 on a Windows 10 with MikTeX and it looks good. The PDF is valid and the figures (PlantUML with filter, and SVG) are showing nicely (although I haven't validated the entire book). Thank you for making this fix.

I did notice that some PDF files (probably from SVG diagrams) get created and remain in my working repo. They show up as Untracked in git. It's my feeling that they should be cleaned up, if they were only created as a conversion step to integrate into the final PDF. 
Another file that's generated and remains (Untracked) is `index.fdb_latexmk` (but the other `index.*` created by the latexmk step seem to be removed).

**Reply by dragonstyle** [reply]
Thank you for trying this out and finding those extraneous artifacts - I will get those cleaned up as I agree they should not be left around!

**Reply by dragonstyle** [reply]
Our next pre-release build will have a change that should address those leftover `pdf` artifacts post SVG conversion. Thanks again for your diligence!

**Reply by fuhrmanator** [reply]
Hello and happy 2023!

Today I tried with 1.3.83 and it's not working anymore. I get the following errors:

```
[WARNING] Citeproc: citation fig-LucidchartUML not found
[WARNING] Citeproc: citation fig-MDD-ErreurDescription not found
[WARNING] Citeproc: citation fig-dssdetails not found
running latexmk - 1
  Latexmk: File-database 'index.fdb_latexmk' is of incompatible version, 3 v. current version 4
  Rc files read:
    NONE
  Latexmk: Run number 1 of rule 'pdflatex'
  This is pdfTeX, Version 3.141592653-2.6-1.40.24 (MiKTeX 22.11.25) (preloaded format=pdflatex.fmt)
   restricted \write18 enabled.
  entering extended mode
  Latexmk: If appropriate, the -f option can be used to get latexmk
    to try to force complete processing.
  Latexmk: Getting log file 'index.log'
  Latexmk: ============== Inequiv cwd_fls cwd 'C:/Users/Me/Documents/GitHub/log210-ndc-quarto' '/c/Users/Me/Documents/GitHub/log210-ndc-quarto'
  Collected error summary (may duplicate other messages):
    pdflatex: Command for 'pdflatex' gave return code 1
        Refer to 'index.log' and/or above output for details



compilation failed- missing packages (automatic installed disabled)
LaTeX Error: Unknown graphics extension: .svg.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...

l.610 ...b55fa4339a2a8e3324e11430a90a75be58a4.svg}


see C:\Users\Me\Documents\GitHub\log210-ndc-quarto\index.log for more information.
```

**Reply by fuhrmanator** [reply]
Background: I can't use the latest 1.2 (because of the blank pdf bug with latexmx), but want to use the shortcodes for font awesome (1.2.263 I think), so I'm forced to go to 1.3. 

I rolled back to 1.3.24 and it's working again (apart from the leftover .pdfs and other files that aren't cleaned up, which I can live with). 

I'm trying to get a release of the book project this week, so I will have more time to figure out which version after 1.3.24 has the change that broke the .svg support in my pipeline (plantuml filter).

**Reply by dragonstyle** [reply, marked as answer]
Thanks for finding this- it was a regression in some more recent changes. I've made a fix and am re-running a build (just FYI, understand if you want to stick with the previous version).
